### ISSUE TITLE
SimpleSAML_Error_Exception: Error 8 - Array to string conversion

### ISSUE THREAD

**Issue by jh23453** [privacyidea/simplesamlphp-module-privacyidea#19]
In privacyidea.txt we have the following attributemap:

```
         'attributemap' => array('username' => 'samlLoginName',
         			 'surname' => 'surName',
         			 'givenname' => 'givenName',
         			 'email' => 'emailAddress',
         			 'phone' => 'telePhone',
         			 'mobile' => 'mobilePhone'),
```

Since https://github.com/privacyidea/privacyidea/commit/330591e872646f57e1a78989837106193376d498
the LDAP resolver returns an array of mobile numbers. And we get in the log:

```
Jun 29 22:00:22 simplesamlphp DEBUG [00e11fd73e] privacyidea        key: mobile
Jun 29 22:00:22 simplesamlphp DEBUG [00e11fd73e] privacyidea mapped key: mobilePhone
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] SimpleSAML_Error_Exception: Error 8 - Array to string conversion
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] Backtrace:
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] 4 /usr/share/simplesamlphp/www/_include.php:75 (SimpleSAML_error_handler)
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] 3 /usr/share/simplesamlphp/modules/privacyidea/lib/Auth/Source/privacyidea.php:262 (sspmod_privacyidea_Auth_Source_priv
acyidea::login_chal_resp)
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] 2 /usr/share/simplesamlphp/modules/privacyidea/lib/Auth/Source/privacyidea.php:356 (sspmod_privacyidea_Auth_Source_priv
acyidea::handleLogin)
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] 1 /usr/share/simplesamlphp/modules/privacyidea/www/loginform.php:101 (require)
Jun 29 22:00:22 simplesamlphp ERROR [00e11fd73e] 0 /usr/share/simplesamlphp/www/module.php:127 (N/A)
```

Can we enhance the documentation to either use the number or the first element of the array and avoid the ERROR messages? Since I don't use the field I simply removed the field from the attributemap.

**Comment by jh23453** [comment]
Maybe the change is needed in privacyidea/api/validate.py, line 391?
```
384             # privacyIDEA's own attribute map
385             result_obj["attributes"] = {"username": ui.get("username"),
386                                         "realm": user.realm,
387                                         "resolver": user.resolver,
388                                         "email": ui.get("email"),
389                                         "surname": ui.get("surname"),
390                                         "givenname": ui.get("givenname"),
391                                         "mobile": ui.get("mobile"),
392                                         "phone": ui.get("phone")
393                                         }
```

**Comment by cornelinux** [comment]
The ldap resolver even enhanced more to support even more multi value attributes. So there could be other attributes - not only the mobile number - that would return an array.

So I would prefer to have the simpleSAMLphp decide, if to take the first value if there is an array or to also return the array.

I think the ``attribute_value`` here could be actually an array:
https://github.com/privacyidea/simplesamlphp-module-privacyidea/blob/master/lib/Auth/Source/privacyidea.php#L269

What do you think?

**Comment by jh23453** [comment]
Something like that should work too.

**Comment by cornelinux** [comment]
At the moment the function ``login_chal_resp`` always returns an attribute array, that acutally contains an array in each entry (need to check this with existing attributemap and non-existing attributemap.)

    [ "mobile" => [ "m1", "m2" ],
      "email" => ["e1", "e2"],
      "givenname" => ["hans"]]

The question is, what we want to return as an alternative:

A) The first value of each sub-array as an array

    ["mobile" => ["m1"]]

B) The first value of each sub-array as a string

     ["mobile" => "m1"]

...and if we want to be able to configure the per key or in general.

**Comment by jh23453** [comment]
I've remove mobile numbers from my LDAP (it's not needed here anyway).
So I'd suggest looking for feedback from people that really have these fields.

**Comment by cornelinux** [comment]
@Mipronimo Do you have an idea, how this could be configured?

Usually the attribute map looks like this:

          'attributemap' => array(
          			 'username' => 'samlLoginName',
          			 'surname' => 'surName',
          			 'givenname' => 'givenName',
          			 'email' => 'emailAddress',
          			 'phone' => 'telePhone',
￼          			 'mobile' => 'mobilePhone'
￼          			 ),

I think there are two possibilities:

A) Choose a destinct value:

We could allow for choosing a destinct value: "Pick the first value from a list" or "pick the 3rd value from the list.

B) Concatenate all values to a string.

If there are several values in an attribute like: several mobile numbers, or (more realistic) several groups memberships, we could concatenate them to a comma separated string-list. like

    "value1, value2, value3".

C) Return an array.

Wouldn't it be possible to return in fact an array?

   ["value1", "value2", "value3"]

Or does simplesamlphp require to get a string?

**Comment by cornelinux** [comment]
This was fixed with pull request #25 
All values are returned as array.
